Automattic/wp-calypso
# Checkout: Refactor CountrySpecificPaymentFields to explicitly render each form field

#### Proposed Changes

The `CountrySpecificPaymentFields` component, used to collect additional data for the Ebanx credit card and Netbanking payment methods, uses a very dynamic method to create its form fields. It calls `React.createElement()` directly (via a `createField()` helper method) and provides certain props derived from defaults and a standardized get/set interface. 

https://github.com/Automattic/wp-calypso/blob/b472d87a530486b28f61a9944cb8b540a14478fc/client/my-sites/checkout/checkout/country-specific-payment-fields.jsx#L133-L139

https://github.com/Automattic/wp-calypso/blob/b472d87a530486b28f61a9944cb8b540a14478fc/client/my-sites/checkout/checkout/country-specific-payment-fields.jsx#L66-L85

It _looks_ very elegant. However, the interface (what kind of component it can work with) is implicit and not documented, and doesn't apply to all form elements (eg: the phone input and the state input use considerably different props). 

This requires several special cases and non-obvious prop overrides. In addition, the abstraction of calling `createElement()` obfuscates a lot of what is happening and can be hard to debug.

This PR refactors the component to convert it to TypeScript and to render each of its form elements explicitly (with a custom wrapper to deduplicate behavior) instead of using the `createField()`/`createElement()` abstraction (trying to make the abstraction work with TypeScript was extremely difficult and in this case I think indicates the risky nature of the abstraction).

https://github.com/Automattic/wp-calypso/blob/4c3d9ea9b076b665ab14f6896e3de9f566248743/client/my-sites/checkout/composite-checkout/components/country-specific-payment-fields.tsx#L305-L318

Example screenshot of Ebanx form using `CountrySpecificPaymentFields`; the form should be identical before and after this change.

<img width="387" alt="Screen Shot 2022-09-07 at 5 44 45 PM" src="https://user-images.githubusercontent.com/2036909/188988627-cfc0a003-e2ac-4bcd-9ebc-17add26c6ffe.png">

#### Testing Instructions

- Set your user account currency to BRL.
- Visit `/plans`, add a product to your cart, and visit checkout.
- Edit the billing information step if it is not already active.
- Set the country to "Brazil", postal code to `68900-000`, and press "Continue".
- In the payment method step, select "Credit or debit card".
- Verify that you see the "Taxpayer Identification Number" field and all the other additional Ebanx fields (see screenshot above).
- Use an [ebanx test card](https://developer.ebanx.com/docs/resources/testCreditCards/) like `4111111111111111` and [test customer info](https://developer.ebanx.com/docs/resources/testCustomerData/) to fill out the Ebanx form.
- Submit the purchase and verify that the payment succeeds.

There's no need to test Netbanking as it is currently disabled on the back-end. However, with some hacking I was able to make it work for testing. Here's what it looks like before and after this change:

<img width="489" alt="netbanking-after" src="https://user-images.githubusercontent.com/2036909/189172528-63cd5140-8423-4a95-b803-6c5f987ae873.png">
